mlos_bench/mlos_bench/tests/storage/trial_telemetry_test.py

```diff
@@ -0,0 +1,73 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Unit tests for saving and restoring the telemetry data.
+"""
+from datetime import datetime, timedelta
+from typing import Any, List, Optional, Tuple
+
+import pytest
+
+from mlos_bench.environments.status import Status
+from mlos_bench.tunables.tunable_groups import TunableGroups
+from mlos_bench.storage.base_storage import Storage
+
+# pylint: disable=redefined-outer-name
+
+
+@pytest.fixture
+def telemetry_data() -> List[Tuple[datetime, str, Any]]:
+    """
+    Mock telemetry data for the trial.
+
+    Returns
+    -------
+    List[Tuple[datetime, str, str]]
+        A list of (timestamp, metric_id, metric_value)
+    """
+    timestamp1 = datetime.now()
+    timestamp2 = timestamp1 + timedelta(seconds=1)
+    return sorted([
+        (timestamp1, "cpu_load", 10.1),
+        (timestamp1, "memory", 20),
+        (timestamp1, "setup", "prod"),
+        (timestamp2, "cpu_load", 30.1),
+        (timestamp2, "memory", 40),
+        (timestamp2, "setup", "prod"),
+    ])
+
+
+def _telemetry_str(data: List[Tuple[datetime, str, Any]]
+                   ) -> List[Tuple[datetime, str, Optional[str]]]:
+    """
+    Convert telemetry values to strings.
+    """
+    return [(ts, key, None if val is None else str(val)) for (ts, key, val) in data]
+
+
+def test_update_telemetry(exp_storage_memory_sql: Storage.Experiment,
+                          tunable_groups: TunableGroups,
+                          telemetry_data: List[Tuple[datetime, str, Any]]) -> None:
+    """
+    Make sure update_telemetry() and load_telemetry() methods work.
+    """
+    trial = exp_storage_memory_sql.new_trial(tunable_groups)
+    assert exp_storage_memory_sql.load_telemetry(trial.trial_id) == []
+
+    trial.update_telemetry(Status.RUNNING, telemetry_data)
+    assert exp_storage_memory_sql.load_telemetry(trial.trial_id) == _telemetry_str(telemetry_data)
+
+
+def test_update_telemetry_twice(exp_storage_memory_sql: Storage.Experiment,
```

Comment:
Add a test for `IntegrityConstraint` handling when metrics are attempted to be inserted twice?

Reply:
oh, we actually swallow that exception and only print a warning. We want the `.update_telemetry()` method to be idempotent. It is totally fine to retrieve the same telemetry data (that is, unique pairs of [timestamp, metric_id]) twice - e.g., when on the remote VM a script keeps appending this data to a file and we repeatedly download the same file and ingest it

Reply:
Yeah, I saw that.  We should add a test for that was my suggestion.

Reply:
Looks like right now we do it in a loop, which is ineffecient but the only correct way to handle the scenario you describe.
So, to prevent some "smart" person (i.e., me) from trying to turn that into a more effecient bulk insert in the future, where leading row repeats would fail, and cause subsequent rows to also fail even if they're new, we should check that for this handling logic.

Reply:
oh, I finally understand what you mean. I did some research on SQLAlchemy - there is a function `Insert.on_conflict_do_nothing()`, but it is only available for a few dialects (and MySQL is _not_ one of them) ☹️ 
I'll play with it a bit more later - apparently, there is an upsert functionality in a higher-level Session API; I'll see if we can use it in our code

Reply:
I cannot figure it out now. I've added a comment in `.update_telemetry()` explaining the decision; I'll replace it with a better solution later (most likely, explicitly check if current dialect supports `on_conflict_do_nothing` and use bulk insert if it does).